File: split/doregistercommand.pas
procedure doregistercommand(a: smallnumber);
label
  40, 10;
var
  l, q, r, s: halfword;
  p: 0..3;
begin
  q := curcmd;
  {1237:}
begin
    if q <> 89 then
    begin
      getxtoken;
      if (curcmd >= 73) and (curcmd <= 76) then
      begin
        l := curchr;
        p := curcmd - 73;
        goto 40;
      end;
      if curcmd <> 89 then
      begin
        begin
          if interaction = 3 then;
          printnl(262);
          print(685);
        end;
        printcmdchr(curcmd, curchr);
        print(686);
        printcmdchr(q, 0);
        begin
          helpptr := 1;
          helpline[0] := 1209;
        end;
        error;
        goto 10;
      end;
    end;
    p := curchr;
    scaneightbitint;
    case p of
      0: l := curval + 5318;
      1: l := curval + 5851;
      2: l := curval + 2900;
      3: l := curval + 3156;
    end;
  end;
  40:
  {:1237}
;
  if q = 89 then
    scanoptionalequals
  else if scankeyword(1205) then;
  aritherror := false;
  if q < 91 then
    {1238:}
if p < 2 then
    begin
      if p = 0 then
        scanint
      else
        scandimen(
          false, false, false);
      if q = 90 then
        curval := curval + eqtb[l].int;
    end
    else
    begin
      scanglue(p);
      if q = 90 then{1239:}

      begin
        q := newspec(curval);
        r := eqtb[l].hh.rh;
        deleteglueref(curval);
        mem[q + 1].int := mem[q + 1].int + mem[r + 1].int;
        if mem[q + 2].int = 0 then
          mem[q].hh.b0 := 0;
        if mem[q].hh.b0 = mem[r].hh.b0 then
          mem[q + 2].int := mem[q + 2].int + mem[r + 2].int
        else if (mem[q].hh.b0 < mem[r].hh.b0) and (mem[r + 2].int <> 0) then
        begin
          mem
            [q + 2].int := mem[r + 2].int;
          mem[q].hh.b0 := mem[r].hh.b0;
        end;
        if mem[q + 3].int = 0 then
          mem[q].hh.b1 := 0;
        if mem[q].hh.b1 = mem[r].hh.b1 then
          mem[q + 3].int := mem[q + 3].int + mem[r + 3].int
        else if (mem[q].hh.b1 < mem[r].hh.b1) and (mem[r + 3].int <> 0) then
        begin
          mem
            [q + 3].int := mem[r + 3].int;
          mem[q].hh.b1 := mem[r].hh.b1;
        end;
        curval := q;
      end{:1239}
;
    end{:1238}

  else{1240:}

  begin
    scanint;
    if p < 2 then
      if q = 91 then
        if p = 0 then
          curval := multandadd(eqtb[l].int, curval, 0, 2147483647)
        else
          curval := multandadd(eqtb[l].int, curval, 0, 1073741823)
      else
        curval := xovern(eqtb[l].int, curval)
    else
    begin
      s := eqtb[l].hh.rh;
      r := newspec(s);
      if q = 91 then
      begin
        mem[r + 1].int := multandadd(mem[s + 1].int, curval, 0, 1073741823);
        mem[r + 2].int := multandadd(mem[s + 2].int, curval, 0, 1073741823);
        mem[r + 3].int := multandadd(mem[s + 3].int, curval, 0, 1073741823);
      end
      else
      begin
        mem[r + 1].int := xovern(mem[s + 1].int, curval);
        mem[r + 2].int := xovern(mem[s + 2].int, curval);
        mem[r + 3].int := xovern(mem[s + 3].int, curval);
      end;
      curval := r;
    end;
  end{:1240}
;
  if aritherror then
  begin
    begin
      if interaction = 3 then;
      printnl(262);
      print(1206);
    end;
    begin
      helpptr := 2;
      helpline[1] := 1207;
      helpline[0] := 1208;
    end;
    if p >= 2 then
      deleteglueref(curval);
    error;
    goto 10;
  end;
  if p < 2 then
    if (a >= 4) then
      geqworddefine(l, curval)
    else
      eqworddefine(l,
        curval)
  else
  begin
    trapzeroglue;
    if (a >= 4) then
      geqdefine(l, 117, curval)
    else
      eqdefine(l, 117, curval);
  end;
  10: ;
end;

File: cpp/doregistercommand.cpp
#include "doregistercommand.h"
#include "getxtoken.h"
#include "printnl.h"
#include "print.h"
#include "printcmdchr.h"
#include "error.h"
#include "scaneightbitint.h"
#include "scanoptionalequals.h"
#include "scankeyword.h"
#include "scanint.h"
#include "scandimen.h"
#include "scanglue.h"
#include "newspec.h"
#include "deleteglueref.h"
#include "multandadd.h"
#include "xovern.h"
#include "trapzeroglue.h"
#include "geqworddefine.h"
#include "eqworddefine.h"
#include "geqdefine.h"
#include "eqdefine.h"

void doregistercommand(smallnumber a)
{
	halfword l, q, r, s;
	char  p;
	q = curcmd;
	do 
	{
		if (q != register_)
		{
			getxtoken();
			if (curcmd >= assign_int && curcmd <= assign_mu_glue) // assign_int, assign_dimen, assign_int, assign_mu_glue
			{
				l = curchr;
				p = curcmd-assign_int;
				continue;
			}
			if (curcmd != register_)
			{
				printnl(262); //! 
				print(685); //You can't use `
				printcmdchr(curcmd, curchr);
				print(686); //' after 
				printcmdchr(q, 0);
				helpptr = 1;
				helpline[0] = 1209; //I'm forgetting what you said and not changing anything.
				error();
				return;
			}
		}
		p = curchr;
		scaneightbitint();
		switch (p)
		{
			case 0: 
				l = curval+count_base;
				break;
			case 1: 
				l = curval+scaled_base;
				break;
			case 2: 
				l = curval+skip_base;
				break;
			case 3: 
				l = curval+mu_skip_base;
				break;
		}
	} while (false);
	if (q == register_)
		scanoptionalequals();
	else 
		if (scankeyword(1205)) //by
			aritherror = false;
	if (q < multiply)
		if (p < 2)
		{
			if (p == 0)
				scanint();
			else
				scandimen(false, false, false);
			if (q == advance)
				curval += eqtb[l].int_;
		}
		else
		{
			scanglue(p);
			if (q == advance)
			{
				q = newspec(curval);
				r = equiv(l);
				deleteglueref(curval);
				width(q) += width(r);
				if (stretch(q) == 0)
					stretch_order(q) = 0;
				if (stretch_order(q) == stretch_order(r))
					stretch(q) += stretch(r);
				else 
					if (stretch_order(q) < stretch_order(r) && stretch(r))
					{
						stretch(q) = stretch(r);
						stretch_order(q) = stretch_order(r);
					}
				if (shrink(q) == 0)
					shrink_order(q) = 0;
				if (shrink_order(q) == shrink_order(r))
					shrink(q) += shrink(r);
				else 
					if (shrink_order(q) < shrink_order(r) && shrink(r))
					{
						shrink(q) = shrink(r);
						shrink_order(q) = shrink_order(r);
					}
				curval = q;
			}
		}
	else
	{
		scanint();
		if (p < 2)
			if (q == multiply)
				if (p == 0)
					curval = multandadd(eqtb[l].int_, curval, 0, infinity);
				else
					curval = multandadd(eqtb[l].int_, curval, 0, max_dimen);
			else
				curval = xovern(eqtb[l].int_, curval);
		else
		{
			s = equiv(l);
			r = newspec(s);
			if (q == multiply)
			{
				width(r) = multandadd(width(s), curval, 0, max_dimen);
				depth(r) = multandadd(depth(s), curval, 0, max_dimen);
				height(r) = multandadd(height(s), curval, 0, max_dimen);
			}
			else
			{
				width(r) = xovern(width(s), curval);
				depth(r) = xovern(depth(s), curval);
				height(r) = xovern(height(s), curval);
			}
			curval = r;
		}
	}
	if (aritherror)
	{
		printnl(262); //! 
		print(1206); //Arithmetic overflow
		helpptr = 2;
		helpline[1] = 1207; //I can't carry out that multiplication or division,
		helpline[0] = 1208; //since the result is out of range.
		if (p >= 2)
			deleteglueref(curval);
		error();
		return;
	}
	if (p < 2)
		if (a >= 4)
			geqworddefine(l, curval);
		else
		eqworddefine(l, curval);
	else
	{
		trapzeroglue();
		if (a >= 4)
			geqdefine(l, 117, curval);
		else
			eqdefine(l, 117, curval);
	}
}
